src/components/formio/fieldset.tsx

```diff
@@ -0,0 +1,64 @@
+import {css} from '@emotion/css';
+import clsx from 'clsx';
+import React from 'react';
+
+import {AnyComponentSchema} from '@/types';
+import {useValidationErrors} from '@/utils/errors';
+import {ErrorList} from '@/utils/errors';
+
+import Tooltip from './tooltip';
+
+export interface FieldsetProps {
+  // XXX: eventually (most) of these literals will be included in AnyComponentType
+  type: AnyComponentSchema['type'] | 'datagrid' | 'datamap' | 'select' | 'columns' | 'textarea';
+  field?: string;
+  required?: boolean;
+  label?: React.ReactNode;
+  tooltip?: string;
+  htmlId?: string;
+  children: React.ReactNode;
+}
+
+// Fix the overlapping icons/text when the error icon is shown.
+// XXX: once we've moved away from bootstrap/formio 'component library', this fix and
+// @emotion/css can be removed again.
+const PAD_ERROR_ICON = css`
+  .form-control.is-invalid {
+    padding-inline-end: calc(1.5em + 0.75rem);
+  }
+`;
+
+const Fieldset: React.FC<FieldsetProps> = ({
+  type,
+  field = '',
+  required = false,
+  label,
+  tooltip = '',
+  children,
+}) => {
+  const {errors} = useValidationErrors(field);
+  const className = clsx('form-group', 'has-feedback', 'formio-component', PAD_ERROR_ICON, {
+    [`formio-component-${type}`]: type,
```

Comment:
do we need this class name? there's definitely no built in formio styles targetting this since it's a custom component